lib/token.h

```diff
@@ -187,6 +187,30 @@ class CPPCHECKLIB Token {
     explicit Token(TokensFrontBack *tokensFrontBack = nullptr);
     ~Token();
 
+    struct TokenEnumeration
+    {
+        Token const* mt;
+        TokenEnumeration(Token const* t) : mt(t) {}
+
+        struct TokenIterator
+        {
+            const Token* mt;
+            TokenIterator(const Token* t) : mt(t) {}
+            TokenIterator& operator++() { mt = mt->next(); return *this; }
+            bool operator==(const TokenIterator& b) { return mt == b.mt; }
+            bool operator!=(const TokenIterator& b) { return mt != b.mt; }
+            const Token* operator*() { return mt; }
+        };
+
+        TokenIterator begin() { return TokenIterator(mt); }
+        TokenIterator end() { return TokenIterator(mt->mTokensFrontBack->back); }
```

Comment:
Shouldn't the end token be nullptr if we want to loop through all tokens?

As @pfultz2 says, it would be very useful with partial ranges.

We have such code in many places:

    for (const Token *tok = someScope->bodyStart; tok != someScope->bodyEnd; tok = tok->next())

I would like to replace that with something like:

    for (const Token *tok: someScope->bodyTokens())

Reply:
I think you're probably right and that this contains a bug. I'll be adding some unit tests which hopefully will identify this bug.

Reply:
I've added tests and this was indeed a bug

Reply:
example of a partial range can be seen in the test ```partialEnumeration```. It looks like this
```
        const Token* expected = var.tokens()->tokAt(4);
        const Token* end = var.tokens()->tokAt(10);
        for (auto t : TokenRange<const Token>{ expected, end }) {
```